text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,553 @@
+  * [Falling back to explicit contexts](#Falling-back-to-explicit-contexts)
+* [Internal details](#Internal-details)
+* [FAQ](#faq)
+
+Refactor OpenTelemetry into a set of separate cross-cutting concerns which 
+operate on a shared context propagation mechanism.
+
+# Motivation
+
+This RFC addresses the following topics:
+
+**Separatation of concerns**  
+* Remove the Tracer dependency from context propagation mechanisms.
+* Handle user data (Correlations) and observability data (SpanContext, etc) 
+  separately.  
+
+**Extensibility**
+* Allow developers to create new applications for context propagation. For 
+  example: A/B testing, encrypted or authenticated data, and new, experimental 
+  forms of observability.
+
+## Explanation
+
+
+# OpenTelemetry layered architecture
+
+The design of OpenTelemetry is based on the priciples of [aspect-oriented 
+programming](https://en.wikipedia.org/wiki/Aspect-oriented_programming), 
+adopted to the needs of distributed systems.
+
+OpenTelemetry is separated into two layers. The top layer contains a set of 
+independent **cross-cutting concerns**, which intertwine with a program's 
+application logic, and cannot be cleanly encapsulated. All concerns share an 
+underlying distributed **context propagation** layer, for storing state and 
+accessing data across the lifespan of a distributed transaction.
+
+
+# Cross-Cutting Concerns
+
+## Observability API
+Distributed tracing is one example of a cross-cutting concern. Tracing code is 
+interleaved with regular code, and ties together independent code modules which 
+would otherwise remain encapsulated. Tracing is also distributed, and requires 
+non-local, transaction-level context propagation in order to execute correctly.
+
+The various observability APIs are not described here directly. However, in this new 
+design, all observability APIs would be modified to make use of the generalized 
+context propagation mechanism described below, rather than the tracing-specific 
+propagation system it uses today.
+
+Note that OpenTelemetry APIs calls should *always* be given access to the entire 
+context object, and never just a subset of the context, such as the value in a 
+single key. This allows the SDK to make improvements and leverage additional 
+data that may be available, without changes to all of the call sites.
+
+## Correlations API
+
+In addition to trace propagation, OpenTelemetry provides a simple mechanism for 
+propagating arbitrary transaction-level data, called Correlations. Correlations 
+implements an extremely basic form of cross-cutting concern - a transaction-level 
+key-value store. Values placed in the 
+
+Event correlation is a mechanism for indexing events in one service with 
+attributes provided by a prior service in the same transaction. This helps to 
+establish a causal relationship between these events. For example, determining 
+that a particular browser version is associated with a failure in an image 
+processing service.
```

Comment:
Don't like this paragraph.  It emphasizes the "observability" function of Correlations. Baggage is a mechanism of associating metadata with the whole distributed transaction, a purpose much broader than just correlating events.

Reply:
I am trying to split the difference: allowing Correlations to be used for backwards compatibility with OpenTracing Baggage, while removing the concept of Baggage from OpenTelemetry. 

Because the Correlations API includes getters, this concern will never have any 
tricky optimizations related to sending data out of band. So it is safe for use 
as a backwards-compatibility mechanism for any OpenTracing baggage which already exists. I added MUST clauses to emphasize that this is safe.

My hope is that we can remove the concept of general-purpose user data as an intended use case for Correlations, and at the same time demonstrate how the prior use cases for Baggage could be easily rebuilt outside of OpenTelemetry using the same context propagation primitives - which themselves could be moved out to a neutral project. That way, these other concerns do not need to "camp" on the observability system.

In the short term, it is still possible to continue to use Correlations like it was Baggage, given that it is safe. But the long term, the explicit intention is that these are separate concerns.